packages/core/src/saml-applications/routes/utils.ts

```diff
@@ -28,8 +29,366 @@ import {
+        {
+          Location: singleSignOnUrl,
+          Binding: BindingType.Redirect,
+        },
+        {
+          Location: singleSignOnUrl,
+          Binding: BindingType.Post,
+        },
+      ],
+      privateKey,
+      isAssertionEncrypted: false,
+      loginResponseTemplate: {
+        context: samlLogInResponseTemplate,
+        attributes: [
+          {
+            name: 'email',
+            valueTag: 'email',
+            nameFormat: samlAttributeNameFormatBasic,
+            valueXsiType: samlValueXmlnsXsi.string,
+          },
+          {
+            name: 'name',
+            valueTag: 'name',
+            nameFormat: samlAttributeNameFormatBasic,
+            valueXsiType: samlValueXmlnsXsi.string,
+          },
+        ],
+      },
+      nameIDFormat: [
+        saml.Constants.namespace.format.emailAddress,
+        saml.Constants.namespace.format.persistent,
+      ],
+    });
+
+    // eslint-disable-next-line new-cap
+    const sp = saml.ServiceProvider({
+      entityID: spEntityId,
+      assertionConsumerService: [
+        {
+          Binding: acsUrl.binding,
+          Location: acsUrl.url,
+        },
+      ],
+      signingCert: certificate,
+      authnRequestsSigned: idp.entityMeta.isWantAuthnRequestsSigned(),
+      allowCreate: false,
+    });
+
+    // Used to check whether xml content is valid in format.
+    saml.setSchemaValidator({
+      validate: async (xmlContent: string) => {
+        try {
+          XMLValidator.validate(xmlContent, {
+            allowBooleanAttributes: true,
+          });
+
+          return true;
+        } catch {
+          return false;
+        }
+      },
+    });
+
+    return { idp, sp };
+  };
+
+  public details: ValidSamlApplicationDetails;
+
+  protected idp: saml.IdentityProviderInstance;
+  protected sp: saml.ServiceProviderInstance;
+
+  protected tenantEndpoint: URL;
+  protected oidcConfig?: CamelCaseKeys<OidcConfigResponse>;
+
+  constructor(
+    details: SamlApplicationDetails,
+    tenantId: string,
+    protected samlApplicationId: string,
+    protected issuer: string
+  ) {
+    this.details = SamlApplication.validateSamlApplicationDetails(details);
+
+    this.tenantEndpoint = getTenantEndpoint(tenantId, EnvSet.values);
+    this.samlApplicationId = samlApplicationId;
+
+    const { idp, sp } = SamlApplication.getSamlIdpAndSp({
+      idp: {
+        entityId: buildSamlIdentityProviderEntityId(this.tenantEndpoint, this.samlApplicationId),
+        privateKey: this.details.privateKey,
+        certificate: this.details.certificate,
+        singleSignOnUrl: buildSingleSignOnUrl(this.tenantEndpoint, this.samlApplicationId),
+      },
+      sp: {
+        entityId: this.details.entityId,
+        acsUrl: this.details.acsUrl,
+      },
+    });
+    this.idp = idp;
+    this.sp = sp;
+  }
+
+  public getIdPMetadata() {
+    return this.idp.getMetadata();
+  }
+
+  public getIdPCertificate() {
+    return this.details.certificate;
+  }
```

Comment:
Can these two methods be removed if the `idp` and `details` are public properties?  Or you can make them as getter methods.